torch/distributed/distributed_c10d.py

```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
     )
 
 
-def _get_group_rank(group: ProcessGroup, rank):
+def get_local_rank(group: ProcessGroup, global_rank: int) -> int:
     """
-    Helper that gets a given group's local rank in the group from a given global
-    rank.
+    Translate a global rank into a local rank.
+
+    ``global_rank`` must be part of ``group`` otherwise this raises RuntimeError.
+
+    Args:
+        group (ProcessGroup): ProcessGroup to find the relative rank.
+        global_rank (int): Global rank to query.
+
+    Returns:
+        Local rank of ``global_rank`` relative to ``group``
+
+    N.B. calling this function on the default process group returns identity
     """
     if group is GroupMember.WORLD:
-        raise RuntimeError(
-            "group.WORLD does not have local rank to global " "rank mapping"
-        )
+        return global_rank
     if group not in _pg_group_ranks:
-        raise RuntimeError("The given group does not exist")
-    try:
-        group_rank = _pg_group_ranks[group][rank]
-    except KeyError:
-        raise RuntimeError(
-            f"The global rank {rank} is not part of the group {group}"
-        ) from None
-    return group_rank
+        raise RuntimeError("Group {group} is not registered with torch.distributed")
+    group_ranks = _pg_group_ranks[group]
+    if global_rank not in group_ranks:
+        raise RuntimeError(f"Global rank {global_rank} is not part of group {group}")
 
+    return group_ranks[global_rank]
 
-def _get_global_rank(group, group_rank):
+def get_global_rank(group: ProcessGroup, local_rank: int) -> int:
     """
-    Helper that gets a given group's global rank from a given local rank in the
-    group.
+    Translate a local rank into a global rank.
+
+    ``local_rank`` must be part of `group` otherwise this raises RuntimeError.
+
+    Args:
+        group (ProcessGroup): ProcessGroup to find the global rank from.
+        local_rank (int): Local rank to query.
+
+    Returns:
+        Global rank of ``local_rank`` relative to ``group``
+
+    N.B. calling this function on the default process group returns identity
     """
     if group is GroupMember.WORLD:
-        raise RuntimeError(
-            "group.WORLD does not have local rank to global " "rank mapping"
-        )
-    group_rank_map = _pg_group_ranks[group]
-    for rank, grp_rank in group_rank_map.items():
-        if grp_rank == group_rank:
+        return local_rank
+    if group not in _pg_group_ranks:
+        raise RuntimeError("Group {group} is not registered with torch.distributed")
+    for rank, grp_rank in _pg_group_ranks[group].items():
+        if grp_rank == local_rank:
             return rank
-    raise RuntimeError("The group rank is not part of the group")
+    raise RuntimeError(f"Global rank {local_rank} is not part of group {group}")
 
+def get_global_ranks(group: ProcessGroup) -> Sequence[int]:
```

Comment:
This is the membership list of all ranks that are part of a PG translated to global ranks. It's the equivalent of the following:

```python
def get_global_ranks(pg):
  return [get_global_rank(pg, i) for i in range(0, pg.get_size())]
```

It essentially recovers the `ranks` argument passed to `new_group`.
Is the naming is too similar to `get_global_rank` that makes it confusing?

Reply:
I see. I think the confusing part is nothing is using this API in this PR just yet. Some documentation would definitely help.

On the other hand, is this actually helpful by itself? What I mean is a user still cannot create a copy of a pg as input without knowing other params like timeout and pg backend type. I feel like in practice this particular call may be only useful for that purpose (for most cases I'd imagine `get_global_rank()` is more often used), and hence it might make more sense to offer a high level api like `clone()` (with signature similar to `new_group()`), unless you plan to introduce `get_timeout(pg)`, `get_backend(pg)` which seem a bit awkward.

Reply:
Based on this discussion, it sounds like https://github.com/pytorch/pytorch/issues/58015 might be useful? i.e. to address the following concern:

> What I mean is a user still cannot create a copy of a pg as input without knowing other params like timeout and pg backend type

Reply:
I believe offering introspection into framework objects to be really important regardless of specific usage.

While this API is not enough to implement `clone_group` it's a step in that direction. I plan to fill in the remaining gaps in follow-up PRs. Maybe I should have made this explicit from the beginning.

Reply:
@rohan-varma my plan is to implement [#81291 ](https://github.com/pytorch/pytorch/issues/81291) design of `new_sub_group` and `clone`.